nncf/experimental/common/pruning/operations.py

```diff
@@ -0,0 +1,637 @@
+
+    @classmethod
+    def get_all_op_aliases(cls) -> List[str]:
+        """
+        :return: list of all aliases of types in metatype
+        """
+        op_types = []
+        for subtype in cls.subtypes:
+            op_types.extend(subtype.get_all_aliases())
+        op_types = list(set(op_types)) + cls.additional_types
+        return op_types
+
+
+class InputPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        node.data['output_mask'] = None
+
+
+class OutputPruningOp(BasePruningOp):
+    @classmethod
+    def accept_pruned_input(cls, node: NNCFNode) -> bool:
+        return True
+
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        node.data['output_mask'] = None
+
+
+class IdentityMaskForwardPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        identity_mask_propagation(node, graph)
+
+
+class ConvolutionPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        output_mask = node.data.get('output_mask', None)
+
+        if is_grouped_conv(node):
+            output_mask = None
+            if is_prunable_depthwise_conv(node):
+                output_mask = input_masks[0]
+
+        node.data['output_mask'] = output_mask
+
+
+class TransposeConvolutionPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        output_mask = node.data.get('output_mask', None)
+
+        # In case of group convs we can't prune by output filters
+        if is_grouped_conv(node):
+            output_mask = None
+            if is_prunable_depthwise_conv(node):
+                output_mask = input_masks[0]
+
+        node.data['output_mask'] = output_mask
+
+
+class LinearPruningOp(BasePruningOp):
+    @classmethod
+    def mask_propagation(cls, node: NNCFNode, graph: NNCFGraph,
+                         tensor_processor: Type[NNCFPruningBaseTensorProcessor]) -> None:
+        input_masks = get_input_masks(node, graph)
+        # length=3 in case of input, and apply_mask for weight and bias. E.g. input: [1,128,4] weight: [4,4] bias: [4]
+        assert len(input_masks) in [1, 2, 3]
+        input_masks = list(filter(lambda mask: mask is not None, input_masks))
+        # single non-empty input
```

Comment:
Could you please describe possible cases in more detail?

Reply:
Which cases do you mean?

Reply:
What cases are covered by expressions
```
 if (len(input_masks) == 1 and input_masks[0] is not None) or (
                len(input_masks) > 1 and not all(input_masks[1:])):
 ```
 or
 ```
 elif len(input_masks) == 2 and all(input_masks):
 ```

Reply:
Refactored this part. More inputs were coming from using compressed graph in the JPQD. 
There were additional inputs from sparsity operations. For mask propagation I use original graph now and there's no such issues.
@vuiseng9 @yujiepan-work

Reply:
Now it's clear, thanks a lot!